lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+      FieldsToTransform.erase(MaxF);
+    }
+  }
+
+  // All the remaining fields are cold with others, create one record for all of
+  // them
+  if (FieldsToTransform.size()) {
+    auto *NewRecord = new SubRecordType;
+    SubRecords.push_back(NewRecord);
+    DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT, dbgs() << "All remaining fields are "
+                                                "cold, create one record for "
+                                                "all of them\n");
+    for (auto &F : FieldsToTransform) {
+      DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                      dbgs() << "Put F" << F + 1 << " into the cold record\n");
+      NewRecord->push_back(F);
+    }
+    FieldsToTransform.clear();
+  }
+  assert(FieldsToTransform.size() == 0);
+
+  // Print suggestions
+  outs() << "Struct definition: " << *StructureType << "\n";
+  for (unsigned i = 0; i < SubRecords.size(); i++) {
+    outs() << "Group #" << i + 1 << ":{ ";
+    for (auto &F : *SubRecords[i]) {
+      outs() << "F" << F + 1 << " ";
+    }
+    outs() << "}\n";
+  }
+}
+
+// Functions of OldFieldReorderAnalyzer. Here "old" just means the algorithm
+// used to reorder fields is derived from prior work in GCC and not necessarily
+// mean it's deprecated or worse compared to the other FieldReorderAnalyzer
+OldFieldReorderAnalyzer::OldFieldReorderAnalyzer(
+    const Module &CM, const StructType *ST, const CloseProximityBuilder *CPB,
+    const DICompositeType *DI)
+    : FieldReorderAnalyzer(CM, ST, CPB, DI, true),
+      DistanceThreshold(StructSplitDistanceThreshold) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Initializing OldFieldReorderAnalyzer of size: "
+                         << NumElements << "\n");
+  // Calculate every CP relations and save it to an array to avoid repeated
+  // calculation
+  CloseProximityRelations.resize(NumElements);
+  double MaxWCP = 0;
+  for (unsigned i = 0; i < NumElements; i++) {
+    CloseProximityRelations[i].resize(NumElements);
+    for (unsigned j = 0; j < NumElements; j++) {
+      CloseProximityRelations[i][j] = calculateCloseProximity(i, j);
+      if (CloseProximityRelations[i][j] > MaxWCP)
+        MaxWCP = CloseProximityRelations[i][j];
+    }
+  }
+  if (MaxWCP < 1e-3) {
+    Eligibility = false;
+    return;
+  }
+  DEBUG(dbgs() << "CPG for field reordering recommendation\n");
+  for (unsigned i = 0; i < NumElements; i++) {
+    DEBUG(dbgs() << "F" << i + 1 << "\t");
+    for (unsigned j = 0; j < NumElements; j++)
+      DEBUG(dbgs() << DEBUG_PRINT_COUNT(CloseProximityRelations[i][j]) << "\t");
+    DEBUG(dbgs() << "\n");
+  }
+}
+
+// Function to calculate CP values. In this algorithm, it interprets (count,
+// distance) as using count as it's CP value, if the distance is within a
+// predefined threshold.
+double
+OldFieldReorderAnalyzer::calculateCloseProximity(FieldNumType Field1,
+                                                 FieldNumType Field2) const {
+  if (Field1 == Field2)
```

Comment:
I designed this function to be a pure virtual function so that each derived class needs to implement it. I agree that the three class might have some overlap in this function but I don't think it's easy to summarize as a general function in the base class. Can I keep it the current way?